File: Linked List/src/19.remove-nth-node-from-end-of-list.js
/*
 * @lc app=leetcode id=19 lang=javascript
 *
 * [19] Remove Nth Node From End of List
 */


// Definition for singly-linked list.
class ListNode {
  constructor(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
  }
}

// @lc code=start
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
  const newHead = new ListNode(0, head);

  let slow = newHead;
  let fast = newHead;
  
  while (n-- > 0) {
    fast = fast.next;
  }
  
  while (fast.next !== null) {
    slow = slow.next;
    fast = fast.next;
  }
  
  slow.next = slow.next?.next ?? null;

  return newHead.next;
};
// @lc code=end

File: Top Interview Questions/src/19.remove-nth-node-from-end-of-list.java
/*
 * @lc app=leetcode id=19 lang=java
 *
 * [19] Remove Nth Node From End of List
 */

// Definition for singly-linked list.
class ListNode {
  int val;
  ListNode next;
  ListNode() {}
  ListNode(int val) { this.val = val; }
  ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
// @lc code=start
class Solution {
  public ListNode removeNthFromEnd(ListNode head, int n) {
    ListNode newHead = new ListNode(0, head);

    ListNode fast = newHead;
    ListNode slow = newHead;

    for (int i = 0; i < n; i++) {
      fast = fast.next;
    }

    while (fast.next != null) {
      fast = fast.next;
      slow = slow.next;
    }

    slow.next = slow.next.next;

    return newHead.next;
  }
}
// @lc code=end

